In [3]:
import polars as pl
import numpy as np
# from utils.process_data import process_item_data, explode_user_interactions #TODO see if I need this
from parquet_data_reader import ParquetDataReader
from models.user_based import CollaborativeRecommender

pl.Config.set_tbl_cols(-1)

ModuleNotFoundError: No module named 'scipy'

## Data import and EDA

In [4]:
data_reader = ParquetDataReader()
articles_df = data_reader.read_data('../data/articles.parquet')
train_behaviors_df = data_reader.read_data('../data/train/behaviors.parquet')
train_history_df = data_reader.read_data('../data/train/history.parquet')
document_vectors_df = data_reader.read_data('../data/document_vector.parquet')


ValueError: Failed to read the Parquet file: The system cannot find the file specified. (os error 2): ../data/document_vector.parquet

We check the size of the data. From the sizes we learn that:
<ol>
  <li>We have 20738 unique articles</li>
  <li>We have 15143 unique users</li>
  <li>We have 232887 interactions in the testset</li>
</ol> 

In [49]:
print("articles_df has the size:         ", articles_df.shape)
print("train_behaviors_df has the size:  ", train_behaviors_df.shape)
print("train_history_df has the size:    ", train_history_df.shape)
print("document_vectors_df has the size: ", document_vectors_df.shape)

articles_df has the size:          (20738, 21)
train_behaviors_df has the size:   (232887, 17)
train_history_df has the size:     (15143, 5)
document_vectors_df has the size:  (125541, 2)


### Table contents

The information on news articles. As we are going to perform user-user CF, this table is not neccesary

In [50]:
articles_df.head()

shape: (5, 21)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┐
│ art ┆ tit ┆ sub ┆ las ┆ pre ┆ bod ┆ pub ┆ ima ┆ art ┆ url ┆ ner ┆ ent ┆ top ┆ cat ┆ sub ┆ cat ┆ tot ┆ tot ┆ tot ┆ sen ┆ sen │
│ icl ┆ le  ┆ tit ┆ t_m ┆ miu ┆ y   ┆ lis ┆ ge_ ┆ icl ┆ --- ┆ _cl ┆ ity ┆ ics ┆ ego ┆ cat ┆ ego ┆ al_ ┆ al_ ┆ al_ ┆ tim ┆ tim │
│ e_i ┆ --- ┆ le  ┆ odi ┆ m   ┆ --- ┆ hed ┆ ids ┆ e_t ┆ str ┆ ust ┆ _gr ┆ --- ┆ ry  ┆ ego ┆ ry_ ┆ inv ┆ pag ┆ rea ┆ ent ┆ ent │
│ d   ┆ str ┆ --- ┆ fie ┆ --- ┆ str ┆ _ti ┆ --- ┆ ype ┆     ┆ ers ┆ oup ┆ lis ┆ --- ┆ ry  ┆ str ┆ iew ┆ evi ┆ d_t ┆ _sc ┆ _la │
│ --- ┆     ┆ str ┆ d_t ┆ boo ┆     ┆ me  ┆ lis ┆ --- ┆     ┆ --- ┆ s   ┆ t[s ┆ i16 ┆ --- ┆ --- ┆ s   ┆ ews ┆ ime ┆ ore ┆ bel │
│ i32 ┆     ┆     ┆ ime ┆ l   ┆     ┆ --- ┆ t[i ┆ str ┆     ┆ lis ┆ --- ┆ tr] ┆     ┆ lis ┆ str ┆ --- ┆ --- ┆ --- ┆ --- ┆ --- │
│     ┆     ┆     ┆ --- ┆     ┆     ┆ dat ┆ 64] ┆     ┆     ┆ t[s ┆ lis ┆     ┆     ┆ t[i ┆     ┆ i32 ┆ i32 ┆ f32 ┆ f32 ┆ str │
│     ┆     ┆     ┆ dat ┆     ┆     ┆ eti ┆     ┆     ┆     ┆ tr] ┆ t[s ┆     ┆     ┆ 16] ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆ eti ┆     ┆     ┆ me[ ┆     ┆     ┆     ┆     ┆ tr] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆ me[ ┆     ┆     ┆ μs] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆ μs] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
│ 300 ┆ Nat ┆ Pol ┆ 202 ┆ fal ┆ Sag ┆ 200 ┆ [31 ┆ art ┆ htt ┆ []  ┆ []  ┆ ["K ┆ 140 ┆ []  ┆ kri ┆ nul ┆ nul ┆ nul ┆ 0.9 ┆ Neg │
│ 135 ┆ asc ┆ iti ┆ 3-0 ┆ se  ┆ en  ┆ 6-0 ┆ 508 ┆ icl ┆ ps: ┆     ┆     ┆ rim ┆     ┆     ┆ mi  ┆ l   ┆ l   ┆ l   ┆ 955 ┆ ati │
│ 3   ┆ ha  ┆ et  ┆ 6-2 ┆     ┆ om  ┆ 8-3 ┆ 50] ┆ e_d ┆ //e ┆     ┆     ┆ ina ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ ve  │
│     ┆ var ┆ fry ┆ 9   ┆     ┆ den ┆ 1   ┆     ┆ efa ┆ kst ┆     ┆     ┆ lit ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ ikk ┆ gte ┆ 06: ┆     ┆ øst ┆ 08: ┆     ┆ ult ┆ rab ┆     ┆     ┆ et" ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ e   ┆ r   ┆ 20: ┆     ┆ ris ┆ 06: ┆     ┆     ┆ lad ┆     ┆     ┆ ,   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ den ┆ nu, ┆ 33  ┆     ┆ ke  ┆ 45  ┆     ┆     ┆ et. ┆     ┆     ┆ "Pe ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ før ┆ at  ┆     ┆     ┆ Nat ┆     ┆     ┆     ┆ dk/ ┆     ┆     ┆ rso ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ ste ┆ Nat ┆     ┆     ┆ asc ┆     ┆     ┆     ┆ kri ┆     ┆     ┆ nfa ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ asc ┆     ┆     ┆ ha… ┆     ┆     ┆     ┆ mi/ ┆     ┆     ┆ rli ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ …   ┆     ┆     ┆     ┆     ┆     ┆     ┆ …   ┆     ┆     ┆ g…  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│ 300 ┆ Kun ┆ Bio ┆ 202 ┆ fal ┆ Vat ┆ 200 ┆ [30 ┆ art ┆ htt ┆ []  ┆ []  ┆ ["U ┆ 414 ┆ [43 ┆ und ┆ nul ┆ nul ┆ nul ┆ 0.8 ┆ Pos │
│ 306 ┆ Sta ┆ gra ┆ 3-0 ┆ se  ┆ ika ┆ 6-0 ┆ 067 ┆ icl ┆ ps: ┆     ┆     ┆ nde ┆     ┆ 3,  ┆ erh ┆ l   ┆ l   ┆ l   ┆ 46  ┆ iti │
│ 5   ┆ r   ┆ fgæ ┆ 6-2 ┆     ┆ net ┆ 5-2 ┆ 12] ┆ e_d ┆ //e ┆     ┆     ┆ rho ┆     ┆ 434 ┆ old ┆     ┆     ┆     ┆     ┆ ve  │
│     ┆ War ┆ nge ┆ 9   ┆     ┆ har ┆ 1   ┆     ┆ efa ┆ kst ┆     ┆     ┆ ldn ┆     ┆ ]   ┆ nin ┆     ┆     ┆     ┆     ┆     │
│     ┆ s   ┆ rne ┆ 06: ┆     ┆ opf ┆ 16: ┆     ┆ ult ┆ rab ┆     ┆     ┆ ing ┆     ┆     ┆ g   ┆     ┆     ┆     ┆     ┆     │
│     ┆ tje ┆ str ┆ 20: ┆     ┆ ord ┆ 57: ┆     ┆     ┆ lad ┆     ┆     ┆ ",  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ nte ┆ ømm ┆ 35  ┆     ┆ ret ┆ 00  ┆     ┆     ┆ et. ┆     ┆     ┆ "Fi ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ mer ┆ er 

Each file consists of seven days of impression logs. The train_behaviors_df table contains all information about interactions between users and items, and can be used as a basis for user-user CF. <strong>Therefore we only need this table</strong>.

In [51]:
train_behaviors_df.head()

shape: (5, 17)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┐
│ imp ┆ art ┆ imp ┆ rea ┆ scr ┆ dev ┆ art ┆ art ┆ use ┆ is_ ┆ gen ┆ pos ┆ age ┆ is_ ┆ ses ┆ nex ┆ nex │
│ res ┆ icl ┆ res ┆ d_t ┆ oll ┆ ice ┆ icl ┆ icl ┆ r_i ┆ sso ┆ der ┆ tco ┆ --- ┆ sub ┆ sio ┆ t_r ┆ t_s │
│ sio ┆ e_i ┆ sio ┆ ime ┆ _pe ┆ _ty ┆ e_i ┆ e_i ┆ d   ┆ _us ┆ --- ┆ de  ┆ i8  ┆ scr ┆ n_i ┆ ead ┆ cro │
│ n_i ┆ d   ┆ n_t ┆ --- ┆ rce ┆ pe  ┆ ds_ ┆ ds_ ┆ --- ┆ er  ┆ i8  ┆ --- ┆     ┆ ibe ┆ d   ┆ _ti ┆ ll_ │
│ d   ┆ --- ┆ ime ┆ f32 ┆ nta ┆ --- ┆ inv ┆ cli ┆ u32 ┆ --- ┆     ┆ i8  ┆     ┆ r   ┆ --- ┆ me  ┆ per │
│ --- ┆ i32 ┆ --- ┆     ┆ ge  ┆ i8  ┆ iew ┆ cke ┆     ┆ boo ┆     ┆     ┆     ┆ --- ┆ u32 ┆ --- ┆ cen │
│ u32 ┆     ┆ dat ┆     ┆ --- ┆     ┆ --- ┆ d   ┆     ┆ l   ┆     ┆     ┆     ┆ boo ┆     ┆ f32 ┆ tag │
│     ┆     ┆ eti ┆     ┆ f32 ┆     ┆ lis ┆ --- ┆     ┆     ┆     ┆     ┆     ┆ l   ┆     ┆     ┆ e   │
│     ┆     ┆ me[ ┆     ┆     ┆     ┆ t[i ┆ lis ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ --- │
│     ┆     ┆ μs] ┆     ┆     ┆     ┆ 32] ┆ t[i ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ f32 │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 32] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
│ 149 ┆ nul ┆ 202 ┆ 13. ┆ nul ┆ 2   ┆ [97 ┆ [97 ┆ 139 ┆ fal ┆ nul ┆ nul ┆ nul ┆ fal ┆ 759 ┆ 7.0 ┆ 22. │
│ 474 ┆ l   ┆ 3-0 ┆ 0   ┆ l   ┆     ┆ 786 ┆ 786 ┆ 836 ┆ se  ┆ l   ┆ l   ┆ l   ┆ se  ┆     ┆     ┆ 0   │
│     ┆     ┆ 5-2 ┆     ┆     ┆     ┆ 23, ┆ 57] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 4   ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 07: ┆     ┆     ┆     ┆ 868 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 47: ┆     ┆     ┆     ┆ 2,  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 53  ┆     ┆     ┆     ┆ …   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 872 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 8]  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│ 150 ┆ nul ┆ 202 ┆ 25. ┆ nul ┆ 2   ┆ [97 ┆ [97 ┆ 143 ┆ fal ┆ nul ┆ nul ┆ nul ┆ fal ┆ 124 ┆ 287 ┆ 100 │
│ 528 ┆ l   ┆ 3-0 ┆ 0   ┆ l   ┆     ┆ 787 ┆ 786 ┆ 471 ┆ se  ┆ l   ┆ l   ┆ l   ┆ se  ┆ 0   ┆ .0  ┆ .0  │
│     ┆     ┆ 5-2 ┆     ┆     ┆     ┆ 18, ┆ 23] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 4   ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 07: ┆     ┆     ┆     ┆ 872 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 33: ┆     ┆     ┆     ┆ 8,  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 25  ┆     ┆     ┆     ┆ …   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 868 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 2]  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│ 153 ┆ 977 ┆ 202 ┆ 78. ┆ 100 ┆ 1   ┆ [97 ┆ [97 ┆ 151 ┆ fal ┆ nul ┆ nul ┆ nul ┆ fal ┆ 197 ┆ 45. ┆ 100 │
│ 068 ┆ 868 ┆ 3-0 ┆ 0   ┆ .0  ┆     ┆ 786 ┆ 786 ┆ 570 ┆ se  ┆ l   ┆ l   ┆ l   ┆ se  ┆ 6   ┆ 0   ┆ .0  │
│     ┆ 2   ┆ 5-2 ┆     ┆     ┆     ┆ 57, ┆ 69] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 4   ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 07: ┆     ┆     ┆     ┆ 866 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 09: ┆     ┆     ┆  

Each file consists of users' click histories collected over 21 days period. This table does contain the same values as the train_behaviours_df, but as that table is easier to work with we will use train_behaviours_df over this one

In [52]:
train_history_df.head()

shape: (5, 5)
┌─────────┬──────────────────────┬─────────────────────┬─────────────────────┬─────────────────────┐
│ user_id ┆ impression_time_fixe ┆ scroll_percentage_f ┆ article_id_fixed    ┆ read_time_fixed     │
│ ---     ┆ d                    ┆ ixed                ┆ ---                 ┆ ---                 │
│ u32     ┆ ---                  ┆ ---                 ┆ list[i32]           ┆ list[f32]           │
│         ┆ list[datetime[μs]]   ┆ list[f32]           ┆                     ┆                     │
╞═════════╪══════════════════════╪═════════════════════╪═════════════════════╪═════════════════════╡
│ 13538   ┆ [2023-04-27          ┆ [100.0, 35.0, …     ┆ [9738663, 9738569,  ┆ [17.0, 12.0, …      │
│         ┆ 10:17:43, 2023-04-…  ┆ 100.0]              ┆ … 9769366]          ┆ 16.0]               │
│ 14241   ┆ [2023-04-27          ┆ [100.0, 46.0, …     ┆ [9738557, 9738528,  ┆ [8.0, 9.0, … 12.0]  │
│         ┆ 09:40:18, 2023-04-…  ┆ 100.0]              ┆ … 9767852]          ┆                     │
│ 20396   ┆ [2023-04-27          ┆ [100.0, 59.0, …     ┆ [9738760, 9738355,  ┆ [49.0, 34.0, … 4.0] │
│         ┆ 12:30:44, 2023-04-…  ┆ 13.0]               ┆ … 9769679]          ┆                     │
│ 34912   ┆ [2023-04-29          ┆ [100.0, 35.0, …     ┆ [9741802, 9741804,  ┆ [153.0, 7.0, … 5.0] │
│         ┆ 07:12:49, 2023-04-…  ┆ 27.0]               ┆ … 9770882]          ┆                     │
│ 37953   ┆ [2023-04-27          ┆ [14.0, 28.0, …      ┆ [9739205, 9739202,  ┆ [4.0, 16.0, … 6.0]  │
│         ┆ 19:17:10, 2023-04-…  ┆ 18.0]               ┆ … 9769306]          ┆                     │
└─────────┴──────────────────────┴─────────────────────┴─────────────────────┴─────────────────────┘

List of vectors for each article. This is used to describe the items. It could be used for item-item CF, but is not relevant to user-user CF.  <strong>This table will therefore not be used</strong>

In [53]:
document_vectors_df.head()

shape: (5, 2)
┌────────────┬─────────────────────────────────┐
│ article_id ┆ document_vector                 │
│ ---        ┆ ---                             │
│ i32        ┆ list[f32]                       │
╞════════════╪═════════════════════════════════╡
│ 3000022    ┆ [0.065424, -0.047425, … 0.0357… │
│ 3000063    ┆ [0.028815, -0.000166, … 0.0271… │
│ 3000613    ┆ [0.037971, 0.033923, … 0.06396… │
│ 3000700    ┆ [0.046524, 0.002913, … 0.02342… │
│ 3000840    ┆ [0.014737, 0.024068, … 0.04599… │
└────────────┴─────────────────────────────────┘

From the analasys we see that we only need train_behaviour_df to perform user-user CF

## Preprocessing

### Remove non-needed values

We see that we have several items that are not required for performing user-user CF

In [54]:
train_behaviors_df.head()

shape: (5, 17)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┐
│ imp ┆ art ┆ imp ┆ rea ┆ scr ┆ dev ┆ art ┆ art ┆ use ┆ is_ ┆ gen ┆ pos ┆ age ┆ is_ ┆ ses ┆ nex ┆ nex │
│ res ┆ icl ┆ res ┆ d_t ┆ oll ┆ ice ┆ icl ┆ icl ┆ r_i ┆ sso ┆ der ┆ tco ┆ --- ┆ sub ┆ sio ┆ t_r ┆ t_s │
│ sio ┆ e_i ┆ sio ┆ ime ┆ _pe ┆ _ty ┆ e_i ┆ e_i ┆ d   ┆ _us ┆ --- ┆ de  ┆ i8  ┆ scr ┆ n_i ┆ ead ┆ cro │
│ n_i ┆ d   ┆ n_t ┆ --- ┆ rce ┆ pe  ┆ ds_ ┆ ds_ ┆ --- ┆ er  ┆ i8  ┆ --- ┆     ┆ ibe ┆ d   ┆ _ti ┆ ll_ │
│ d   ┆ --- ┆ ime ┆ f32 ┆ nta ┆ --- ┆ inv ┆ cli ┆ u32 ┆ --- ┆     ┆ i8  ┆     ┆ r   ┆ --- ┆ me  ┆ per │
│ --- ┆ i32 ┆ --- ┆     ┆ ge  ┆ i8  ┆ iew ┆ cke ┆     ┆ boo ┆     ┆     ┆     ┆ --- ┆ u32 ┆ --- ┆ cen │
│ u32 ┆     ┆ dat ┆     ┆ --- ┆     ┆ --- ┆ d   ┆     ┆ l   ┆     ┆     ┆     ┆ boo ┆     ┆ f32 ┆ tag │
│     ┆     ┆ eti ┆     ┆ f32 ┆     ┆ lis ┆ --- ┆     ┆     ┆     ┆     ┆     ┆ l   ┆     ┆     ┆ e   │
│     ┆     ┆ me[ ┆     ┆     ┆     ┆ t[i ┆ lis ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ --- │
│     ┆     ┆ μs] ┆     ┆     ┆     ┆ 32] ┆ t[i ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ f32 │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 32] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
│ 149 ┆ nul ┆ 202 ┆ 13. ┆ nul ┆ 2   ┆ [97 ┆ [97 ┆ 139 ┆ fal ┆ nul ┆ nul ┆ nul ┆ fal ┆ 759 ┆ 7.0 ┆ 22. │
│ 474 ┆ l   ┆ 3-0 ┆ 0   ┆ l   ┆     ┆ 786 ┆ 786 ┆ 836 ┆ se  ┆ l   ┆ l   ┆ l   ┆ se  ┆     ┆     ┆ 0   │
│     ┆     ┆ 5-2 ┆     ┆     ┆     ┆ 23, ┆ 57] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 4   ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 07: ┆     ┆     ┆     ┆ 868 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 47: ┆     ┆     ┆     ┆ 2,  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 53  ┆     ┆     ┆     ┆ …   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 872 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 8]  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│ 150 ┆ nul ┆ 202 ┆ 25. ┆ nul ┆ 2   ┆ [97 ┆ [97 ┆ 143 ┆ fal ┆ nul ┆ nul ┆ nul ┆ fal ┆ 124 ┆ 287 ┆ 100 │
│ 528 ┆ l   ┆ 3-0 ┆ 0   ┆ l   ┆     ┆ 787 ┆ 786 ┆ 471 ┆ se  ┆ l   ┆ l   ┆ l   ┆ se  ┆ 0   ┆ .0  ┆ .0  │
│     ┆     ┆ 5-2 ┆     ┆     ┆     ┆ 18, ┆ 23] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 4   ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 07: ┆     ┆     ┆     ┆ 872 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 33: ┆     ┆     ┆     ┆ 8,  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 25  ┆     ┆     ┆     ┆ …   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 868 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 2]  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│ 153 ┆ 977 ┆ 202 ┆ 78. ┆ 100 ┆ 1   ┆ [97 ┆ [97 ┆ 151 ┆ fal ┆ nul ┆ nul ┆ nul ┆ fal ┆ 197 ┆ 45. ┆ 100 │
│ 068 ┆ 868 ┆ 3-0 ┆ 0   ┆ .0  ┆     ┆ 786 ┆ 786 ┆ 570 ┆ se  ┆ l   ┆ l   ┆ l   ┆ se  ┆ 6   ┆ 0   ┆ .0  │
│     ┆ 2   ┆ 5-2 ┆     ┆     ┆     ┆ 57, ┆ 69] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 4   ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 07: ┆     ┆     ┆     ┆ 866 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 09: ┆     ┆     ┆  

All information that does not describe a user, or a user-item interaction can therefore be removed

In [55]:
irelevant_columns = ["impression_time", "device_type", "article_ids_inview", "article_ids_clicked", "session_id", "next_read_time", "next_scroll_percentage"]
train_behaviors_df = train_behaviors_df.drop(irelevant_columns)
train_behaviors_df.head()

shape: (5, 10)
┌──────────┬──────────┬──────────┬─────────┬─────────┬─────────┬────────┬─────────┬──────┬─────────┐
│ impressi ┆ article_ ┆ read_tim ┆ scroll_ ┆ user_id ┆ is_sso_ ┆ gender ┆ postcod ┆ age  ┆ is_subs │
│ on_id    ┆ id       ┆ e        ┆ percent ┆ ---     ┆ user    ┆ ---    ┆ e       ┆ ---  ┆ criber  │
│ ---      ┆ ---      ┆ ---      ┆ age     ┆ u32     ┆ ---     ┆ i8     ┆ ---     ┆ i8   ┆ ---     │
│ u32      ┆ i32      ┆ f32      ┆ ---     ┆         ┆ bool    ┆        ┆ i8      ┆      ┆ bool    │
│          ┆          ┆          ┆ f32     ┆         ┆         ┆        ┆         ┆      ┆         │
╞══════════╪══════════╪══════════╪═════════╪═════════╪═════════╪════════╪═════════╪══════╪═════════╡
│ 149474   ┆ null     ┆ 13.0     ┆ null    ┆ 139836  ┆ false   ┆ null   ┆ null    ┆ null ┆ false   │
│ 150528   ┆ null     ┆ 25.0     ┆ null    ┆ 143471  ┆ false   ┆ null   ┆ null    ┆ null ┆ false   │
│ 153068   ┆ 9778682  ┆ 78.0     ┆ 100.0   ┆ 151570  ┆ false   ┆ null   ┆ null    ┆ null ┆ false   │
│ 153070   ┆ 9777492  ┆ 26.0     ┆ 100.0   ┆ 151570  ┆ false   ┆ null   ┆ null    ┆ null ┆ false   │
│ 153071   ┆ 9778623  ┆ 125.0    ┆ 100.0   ┆ 151570  ┆ false   ┆ null   ┆ null    ┆ null ┆ false   │
└──────────┴──────────┴──────────┴─────────┴─────────┴─────────┴────────┴─────────┴──────┴─────────┘

The remaining items are the ones that can be used. But already here we see that we have several features with lacking information. We should therefore treat this

### Account for missing values

We see here that alot of the behaviours contain missing values. The therefore have to either remove or replace the values

In [56]:
print(train_behaviors_df.shape)
train_behaviors_df.null_count()

(232887, 10)


shape: (1, 10)
┌──────────┬─────────┬─────────┬─────────┬─────────┬─────────┬────────┬─────────┬────────┬─────────┐
│ impressi ┆ article ┆ read_ti ┆ scroll_ ┆ user_id ┆ is_sso_ ┆ gender ┆ postcod ┆ age    ┆ is_subs │
│ on_id    ┆ _id     ┆ me      ┆ percent ┆ ---     ┆ user    ┆ ---    ┆ e       ┆ ---    ┆ criber  │
│ ---      ┆ ---     ┆ ---     ┆ age     ┆ u32     ┆ ---     ┆ u32    ┆ ---     ┆ u32    ┆ ---     │
│ u32      ┆ u32     ┆ u32     ┆ ---     ┆         ┆ u32     ┆        ┆ u32     ┆        ┆ u32     │
│          ┆         ┆         ┆ u32     ┆         ┆         ┆        ┆         ┆        ┆         │
╞══════════╪═════════╪═════════╪═════════╪═════════╪═════════╪════════╪═════════╪════════╪═════════╡
│ 0        ┆ 162466  ┆ 0       ┆ 163789  ┆ 0       ┆ 0       ┆ 216668 ┆ 228214  ┆ 226546 ┆ 0       │
└──────────┴─────────┴─────────┴─────────┴─────────┴─────────┴────────┴─────────┴────────┴─────────┘

In [57]:
train_behaviors_df = train_behaviors_df.filter(train_behaviors_df["article_id"].is_not_null())
print(train_behaviors_df.shape)
train_behaviors_df.null_count()

(70421, 10)


shape: (1, 10)
┌──────────┬──────────┬─────────┬─────────┬─────────┬─────────┬────────┬─────────┬───────┬─────────┐
│ impressi ┆ article_ ┆ read_ti ┆ scroll_ ┆ user_id ┆ is_sso_ ┆ gender ┆ postcod ┆ age   ┆ is_subs │
│ on_id    ┆ id       ┆ me      ┆ percent ┆ ---     ┆ user    ┆ ---    ┆ e       ┆ ---   ┆ criber  │
│ ---      ┆ ---      ┆ ---     ┆ age     ┆ u32     ┆ ---     ┆ u32    ┆ ---     ┆ u32   ┆ ---     │
│ u32      ┆ u32      ┆ u32     ┆ ---     ┆         ┆ u32     ┆        ┆ u32     ┆       ┆ u32     │
│          ┆          ┆         ┆ u32     ┆         ┆         ┆        ┆         ┆       ┆         │
╞══════════╪══════════╪═════════╪═════════╪═════════╪═════════╪════════╪═════════╪═══════╪═════════╡
│ 0        ┆ 0        ┆ 0       ┆ 1780    ┆ 0       ┆ 0       ┆ 65481  ┆ 68785   ┆ 67972 ┆ 0       │
└──────────┴──────────┴─────────┴─────────┴─────────┴─────────┴────────┴─────────┴───────┴─────────┘

We see that of 70421, are there between 65-68000 missing values for gender, postcode and age. We therefore remove these as there is no use subsidizing them

In [58]:
train_behaviors_df = train_behaviors_df.drop(["gender", "postcode", "age"])
print(train_behaviors_df.shape)
train_behaviors_df.null_count()

(70421, 7)


shape: (1, 7)
┌───────────────┬────────────┬───────────┬─────────────────┬─────────┬─────────────┬───────────────┐
│ impression_id ┆ article_id ┆ read_time ┆ scroll_percenta ┆ user_id ┆ is_sso_user ┆ is_subscriber │
│ ---           ┆ ---        ┆ ---       ┆ ge              ┆ ---     ┆ ---         ┆ ---           │
│ u32           ┆ u32        ┆ u32       ┆ ---             ┆ u32     ┆ u32         ┆ u32           │
│               ┆            ┆           ┆ u32             ┆         ┆             ┆               │
╞═══════════════╪════════════╪═══════════╪═════════════════╪═════════╪═════════════╪═══════════════╡
│ 0             ┆ 0          ┆ 0         ┆ 1780            ┆ 0       ┆ 0           ┆ 0             │
└───────────────┴────────────┴───────────┴─────────────────┴─────────┴─────────────┴───────────────┘

We still see that 1780/70421 rows are missing a scroll percentage. As this is very low (<3%) we can easily replace this. Intitially we just set scroll to 0

In [59]:
train_behaviors_df = train_behaviors_df.fill_null(strategy="zero")

### Account for multiple instances of the same article and user

By checking rows where the user_id and article_id are the same we see that we have 9855 instances where the user has read the same article multiple times

In [60]:
duplicates = train_behaviors_df.group_by(["article_id", "user_id"]).count().filter(pl.col("count") > 1)

print(duplicates)

shape: (9_855, 3)
┌────────────┬─────────┬───────┐
│ article_id ┆ user_id ┆ count │
│ ---        ┆ ---     ┆ ---   │
│ i32        ┆ u32     ┆ u32   │
╞════════════╪═════════╪═══════╡
│ 9772601    ┆ 1641388 ┆ 3     │
│ 9777767    ┆ 1313057 ┆ 2     │
│ 9775850    ┆ 181466  ┆ 2     │
│ 9775804    ┆ 2054719 ┆ 3     │
│ 9772442    ┆ 34873   ┆ 7     │
│ …          ┆ …       ┆ …     │
│ 9775325    ┆ 1812712 ┆ 2     │
│ 9772923    ┆ 1182750 ┆ 5     │
│ 9769559    ┆ 160687  ┆ 2     │
│ 9774568    ┆ 178179  ┆ 3     │
│ 9774549    ┆ 778021  ┆ 2     │
└────────────┴─────────┴───────┘


C:\Users\chris\AppData\Local\Temp\ipykernel_27152\4233841252.py:1: DeprecationWarning: `GroupBy.count` is deprecated. It has been renamed to `len`.
  duplicates = train_behaviors_df.group_by(["article_id", "user_id"]).count().filter(pl.col("count") > 1)


We see that we need to combine these duplicate rows. We therefore propose that for multiple instances of the same article and user, we combine the readtime and select the largest scroll percentage. This way we can preserve the data without having duplicates

In [61]:
train_behaviors_df = (
    train_behaviors_df
    .group_by(["article_id", "user_id"])
    .agg(
        pl.col("read_time").product().alias("total_readtime"),  # Multiply all readtime values
        pl.col("scroll_percentage").max().alias("max_scroll")  # Select the largest scroll percentage
    )
)

print(train_behaviors_df)

shape: (53_349, 4)
┌────────────┬─────────┬────────────────┬────────────┐
│ article_id ┆ user_id ┆ total_readtime ┆ max_scroll │
│ ---        ┆ ---     ┆ ---            ┆ ---        │
│ i32        ┆ u32     ┆ f32            ┆ f32        │
╞════════════╪═════════╪════════════════╪════════════╡
│ 9773078    ┆ 689703  ┆ 47.0           ┆ 100.0      │
│ 9775621    ┆ 1425490 ┆ 10.0           ┆ 100.0      │
│ 9770491    ┆ 105633  ┆ 58.0           ┆ 100.0      │
│ 9771916    ┆ 259531  ┆ 11.0           ┆ 100.0      │
│ 9776322    ┆ 1852019 ┆ 32.0           ┆ 100.0      │
│ …          ┆ …       ┆ …              ┆ …          │
│ 9772366    ┆ 941168  ┆ 20.0           ┆ 100.0      │
│ 9777529    ┆ 1040564 ┆ 130.0          ┆ 100.0      │
│ 9775277    ┆ 1967618 ┆ 137.0          ┆ 100.0      │
│ 9770288    ┆ 2033514 ┆ 89.0           ┆ 0.0        │
│ 9778351    ┆ 814683  ┆ 12.0           ┆ 100.0      │
└────────────┴─────────┴────────────────┴────────────┘


## Model Fit

In [62]:
recommender = CollaborativeRecommender(train_behaviors_df)
recommender.fit()

{689703: [(731181, np.float64(0.7779284719633448)),
  (721432, np.float64(0.7779238190665176)),
  (1481259, np.float64(0.554784314225774)),
  (1737264, np.float64(0.537904032427563)),
  (2199412, np.float64(0.5153606242697127)),
  (1285159, np.float64(0.5102651755775466)),
  (2474382, np.float64(0.4470256910765976)),
  (1978243, np.float64(0.2659658192898453)),
  (1884362, np.float64(0.22879092115397648)),
  (1922107, np.float64(0.2249489028729259))],
 1425490: [(2116058, np.float64(0.9999996568413011)),
  (327503, np.float64(0.9999996568413011)),
  (1358126, np.float64(0.999999656841301)),
  (1598383, np.float64(0.999999656841301)),
  (1559714, np.float64(0.999999656841301)),
  (74819, np.float64(0.9999959596297535)),
  (316814, np.float64(0.9976952749974431)),
  (2377673, np.float64(0.8783916705232241)),
  (461353, np.float64(0.8497753106708351)),
  (781004, np.float64(0.7399398194790708))],
 105633: [(1629235, np.float64(0.5572246326455471)),
  (1279764, np.float64(0.557224632645547

Of the original 15143 users, only 9194 can be accounted for with the current solution. This should be changed in the future

## Model presentation

In [63]:
for user in [630220, 620796, 1067393, 1726258, 17205]:
    print("reccomended for user ", user, ": ", recommender.recommend_n_articles(user_id=user, n=5))

reccomended for user  630220 :  [9778351, 9772453, 9771554, 9776497, 9775484]
reccomended for user  620796 :  [9771224, 9771113, 9778942, 9775905, 9771796]
reccomended for user  1067393 :  [9771113, 9774789, 9772000, 9671904, 9774142]
reccomended for user  1726258 :  [9771224, 9771113, 9769197, 9770450, 9775701]
reccomended for user  17205 :  [9771125, 9778710, 9779748, 9775256, 9773744]
